src/algebra/field.lean

```diff
@@ -1,45 +1,222 @@
 /-
-Copyright (c) 2017 Johannes Hölzl. All rights reserved.
+Copyright (c) 2014 Robert Lewis. All rights reserved.
 Released under Apache 2.0 license as described in the file LICENSE.
-Authors: Johannes Hölzl, Mario Carneiro
+Authors: Robert Lewis, Leonardo de Moura, Johannes Hölzl, Mario Carneiro
 -/
 import algebra.ring
 import algebra.group_with_zero
 open set
 
+set_option default_priority 100 -- see Note [default priority]
+set_option old_structure_cmd true
+
 universe u
 variables {α : Type u}
 
-@[priority 100] -- see Note [lower instance priority]
-instance division_ring.to_domain [s : division_ring α] : domain α :=
-{ eq_zero_or_eq_zero_of_mul_eq_zero := λ a b h,
-    classical.by_contradiction $ λ hn,
-      division_ring.mul_ne_zero (mt or.inl hn) (mt or.inr hn) h
-  ..s, ..(by apply_instance : semiring α) }
+@[ancestor ring has_inv zero_ne_one_class]
+class division_ring (α : Type u) extends ring α, has_inv α, zero_ne_one_class α :=
+(mul_inv_cancel : ∀ {a : α}, a ≠ 0 → a * a⁻¹ = 1)
+(inv_mul_cancel : ∀ {a : α}, a ≠ 0 → a⁻¹ * a = 1)
+(inv_zero : (0 : α)⁻¹ = 0)
+
+section division_ring
+variables [division_ring α] {a b : α}
+
+protected definition algebra.div (a b : α) : α :=
+a * b⁻¹
+
+instance division_ring_has_div : has_div α :=
+⟨algebra.div⟩
+
+lemma division_def (a b : α) : a / b = a * b⁻¹ :=
+rfl
+
+@[simp] lemma inv_zero : 0⁻¹ = (0:α) :=
+division_ring.inv_zero
+
+@[simp] lemma div_zero (a : α) : a / 0 = (0:α) :=
+calc
+  a / 0 = (a:α) * 0⁻¹ : by rw division_def
+    ... = a * 0       : by rw inv_zero
+    ... = (0:α)       : by rw mul_zero
+
+@[simp] lemma mul_inv_cancel (h : a ≠ 0) : a * a⁻¹ = 1 :=
+division_ring.mul_inv_cancel h
+
+@[simp] lemma inv_mul_cancel (h : a ≠ 0) : a⁻¹ * a = 1 :=
+division_ring.inv_mul_cancel h
+
+@[simp] lemma one_div_eq_inv (a : α) : 1 / a = a⁻¹ := one_mul a⁻¹
+
+@[field_simps] lemma inv_eq_one_div (a : α) : a⁻¹ = 1 / a := by simp
 
 /-- Every division ring is a `group_with_zero`. -/
 @[priority 10] -- see Note [lower instance priority]
-instance division_ring.to_group_with_zero {K : Type*} [division_ring K] :
-  group_with_zero K :=
+instance division_ring.to_group_with_zero :
+  group_with_zero α :=
```

Comment:
I thought we're moving towards calling fields `K` instead of `α`?

Reply:
Is a group with zero a field? One thing I don't like about the new convention is that I am never sure what letter to use for these exotic structures. Anyway we can follow this up with some more cleanup in this vein.

Reply:
No, but a division ring is a kind of field.  A Schief*körper*, to be precise.  So obviously it should get a `K` letter.